Mathlib/Data/Matroid/IndepAxioms.lean

```diff
@@ -0,0 +1,560 @@
+
+    set E₀ := insert e (I₀ ∪ B₀)
+    have hE₀fin : E₀.Finite := (hI₀fin.union hB₀fin).insert e
+
+    -- Extend `B₀` to a maximal independent subset of `I₀ ∪ B₀ + e`
+    obtain ⟨J, ⟨hB₀J, hJ, hJss⟩, hJmax⟩ := Finite.exists_maximal_wrt (f := id)
+      (s := {J | B₀ ⊆ J ∧ Indep J ∧ J ⊆ E₀})
+      (hE₀fin.finite_subsets.subset (by simp))
+      ⟨B₀, Subset.rfl, hB₀, (subset_union_right _ _).trans (subset_insert _ _)⟩
+
+    have heI₀ : e ∉ I₀ := not_mem_subset hI₀I heI
+    have heI₀i : Indep (insert e I₀) := ind_mono hins (insert_subset_insert hI₀I)
+
+    have heJ : e ∉ J := fun heJ ↦ hB₀e (ind_mono hJ <| insert_subset heJ hB₀J)
+
+    have hJfin := hE₀fin.subset hJss
+
+    -- We have `|I₀ + e| ≤ |J|`, since otherwise we could extend the maximal set `J`
+    have hcard : (insert e I₀).ncard ≤ J.ncard
+    · refine not_lt.1 fun hlt ↦ ?_
+      obtain ⟨f, hfI, hfJ, hfi⟩ := ind_aug hJ hJfin heI₀i (hI₀fin.insert e) hlt
+      have hfE₀ : f ∈ E₀ := mem_of_mem_of_subset hfI (insert_subset_insert (subset_union_left _ _))
+      refine hfJ (insert_eq_self.1 <| Eq.symm (hJmax _
+        ⟨hB₀J.trans <| subset_insert _ _,hfi,insert_subset hfE₀ hJss⟩ (subset_insert _ _)))
+
+    -- But this means `|I₀| < |J|`, and extending `I₀` into `J` gives a contradiction
+    rw [ncard_insert_of_not_mem heI₀ hI₀fin, ←Nat.lt_iff_add_one_le] at hcard
+
+    obtain ⟨f, hfJ, hfI₀, hfi⟩ := ind_aug (ind_mono hI hI₀I) hI₀fin hJ hJfin hcard
+    exact hI₀ f ⟨Or.elim (hJss hfJ) (fun hfe ↦ (heJ <| hfe ▸ hfJ).elim) (by aesop), hfI₀⟩ hfi )
+
+    ( by
+      rintro X - I hI hIX
+      have hzorn := zorn_subset_nonempty {Y | Indep Y ∧ I ⊆ Y ∧ Y ⊆ X} ?_ I ⟨hI, Subset.rfl, hIX⟩
+      · obtain ⟨J, hJ, -, hJmax⟩ := hzorn
+        exact ⟨J, hJ, fun K hK hJK ↦ (hJmax K hK hJK).subset⟩
+
+      refine fun Is hIs hchain ⟨K, hK⟩ ↦ ⟨⋃₀ Is, ⟨?_,?_,?_⟩, fun _ ↦ subset_sUnion_of_mem ⟩
+      · refine h_compact _ fun J hJ hJfin ↦ ?_
+        have hchoose : ∀ e, e ∈ J → ∃ I, I ∈ Is ∧ (e : α) ∈ I
+        · exact fun _ he ↦ mem_sUnion.1 <| hJ he
+        choose! f hf using hchoose
+        refine J.eq_empty_or_nonempty.elim (fun hJ ↦ hJ ▸ h_empty) (fun hne ↦ ?_)
+        obtain ⟨x, hxJ, hxmax⟩ := Finite.exists_maximal_wrt f _ hJfin hne
+        refine ind_mono (hIs (hf x hxJ).1).1 fun y hyJ ↦ ?_
+        obtain (hle | hle) := hchain.total (hf _ hxJ).1 (hf _ hyJ).1
+        · rw [hxmax _ hyJ hle]; exact (hf _ hyJ).2
+        exact hle (hf _ hyJ).2
+
+      · exact subset_sUnion_of_subset _ K (hIs hK).2.1 hK
+      exact sUnion_subset fun X hX ↦ (hIs hX).2.2 )
+    h_support
+
+@[simp] theorem ofIndepOfFinitary_apply (E : Set α) (Indep : Set α → Prop)
+    (h_empty : Indep ∅)
+    (ind_mono : ∀ ⦃I J⦄, Indep J → I ⊆ J → Indep I)
+    (ind_aug : ∀ ⦃I J⦄, Indep I → I.Finite → Indep J → J.Finite → I.ncard < J.ncard →
+      ∃ e ∈ J, e ∉ I ∧ Indep (insert e I))
+    (h_compact : ∀ I, (∀ J, J ⊆ I → J.Finite → Indep J) → Indep I)
+    (h_support : ∀ ⦃I⦄, Indep I → I ⊆ E) :
```

Comment:
Like a similar comment, you can drop the types here. One reason to do this is that, in case you ever change the types of the arguments to `Matroi.ofIndepOfFinitary` subtly, then you don't have to go through and fix all these. Plus, it gives these simp lemmas less visual prominence.